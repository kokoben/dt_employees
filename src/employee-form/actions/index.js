import * as actions from './types';

export const addEmployee = fields => ({
  type: actions.EMPLOYEE_ADD,
  fields,
});

export const updateFields = fields => ({
  type: actions.FIELDS_UPDATE,
  fields,
});

export const updateSubmitStatus = submitStatus => ({
  type: actions.SUBMIT_UPDATE,
  submitStatus,
});
