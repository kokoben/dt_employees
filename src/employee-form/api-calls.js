// post employee data.
export const postEmployee = () => (
  'https://dt-interviews.appspot.com/'
);
