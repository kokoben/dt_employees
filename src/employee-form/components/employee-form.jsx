import React from 'react';
import { Form, Input, Button } from 'antd';

const FormItem = Form.Item;

const formItemLayout = {
  labelCol: {
    xs: { span: 24 },
    sm: { span: 6 },
  },
  wrapperCol: {
    xs: { span: 24 },
    sm: { span: 18 },
  },
};

const EmployeeForm = () => {
  return (
    <div style={{ width: '50%' }}>
      <Button
        style={{ float: 'left', margin: '10px' }}
        size="small"
        type="primary"
        icon="left"
      >
        Back
      </Button>
      <h1 style={{ clear: 'left' }}>Add Employee</h1>
      <Form style={{ marginLeft: '10px' }}>
        <FormItem
          label="First Name"
          {...formItemLayout}
        >
          <Input />
        </FormItem>
        <FormItem
          label="Last Name"
          {...formItemLayout}
        >
          <Input />
        </FormItem>
        <FormItem
          label="Job Title"
          {...formItemLayout}
        >
          <Input />
        </FormItem>
        <FormItem
          label="Salary"
          {...formItemLayout}
        >
          <Input />
        </FormItem>
        <FormItem
          label="Department"
          {...formItemLayout}
        >
          <Input />
        </FormItem>
      </Form>
    </div>
  );
};

export default EmployeeForm;
