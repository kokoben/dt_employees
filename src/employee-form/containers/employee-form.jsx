import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import { Form, Input, Button, message } from 'antd';
import { Redirect } from 'react-router';
import { Link } from 'react-router-dom';
import { addEmployee, updateFields } from '../actions';
import { setCursor, setCurrentPage } from '../../employees/actions';

const FormItem = Form.Item;

const formItemLayout = {
  labelCol: {
    xs: { span: 24 },
    sm: { span: 6 },
  },
  wrapperCol: {
    xs: { span: 24 },
    sm: { span: 18 },
  },
};

const tailFormItemLayout = {
  wrapperCol: {
    xs: {
      span: 24,
      offset: 0,
    },
    sm: {
      span: 18,
      offset: 6,
    },
  },
};

const EmployeeForm = Form.create({
  onFieldsChange(props, changedFields) {
    props.onChange(changedFields);
  },
  mapPropsToFields(props) {
    return {
      firstName: Form.createFormField(props.firstName),
      lastName: Form.createFormField(props.lastName),
      jobTitle: Form.createFormField(props.jobTitle),
      salary: Form.createFormField(props.salary),
      department: Form.createFormField(props.department),
    };
  },
})((props) => {
  const { getFieldDecorator } = props.form;

  const handleSubmit = (e) => {
    e.preventDefault();
    props.form.validateFields((err, values) => {
      if (!err) {
        props.onGoodSubmit(values);
        message.success('Employee successfully added!', 5);
        props.statusHandler();
        props.pageChanger(1);
        props.cursorChanger(0);
      }
    });
  };

  return (
    <Form
      style={{ marginLeft: '10px' }}
      onSubmit={handleSubmit}
    >
      <FormItem
        label="First Name"
        {...formItemLayout}
      >
        {getFieldDecorator('firstName', {
          rules: [
            { required: true, message: 'Please enter employee\'s first name.', whitespace: true },
            { pattern: /^[a-z ]+$/i, message: 'May only contain letters.' },
          ],
        })(<Input />)}
      </FormItem>
      <FormItem
        label="Last Name"
        {...formItemLayout}
      >
        {getFieldDecorator('lastName', {
          rules: [
             { required: true, message: 'Please enter employee\'s last name.', whitespace: true },
             { pattern: /^[a-z ]+$/i, message: 'May only contain letters.' },
          ],
        })(<Input />)}
      </FormItem>
      <FormItem
        label="Job Title"
        {...formItemLayout}
      >
        {getFieldDecorator('jobTitle', {
          rules: [
             { required: true, message: 'Please enter employee\'s job title.' },
             { pattern: /^[\w ]+$/g, message: 'May only contain alphanumeric characters.' },
          ],
        })(<Input />)}
      </FormItem>
      <FormItem
        label="Salary"
        {...formItemLayout}
      >
        {getFieldDecorator('salary', {
          rules: [
          { required: true, message: 'Please enter employee\'s salary.', whitespace: true },
          {
            pattern: /^[0-9]+(\.[0-9][0-9])?$/,
            message: 'Invalid amount. Must in the format: (dollar amount).(cents) Example: 19899.52',
          },
          ],
        })(<Input prefix="$" />)}
      </FormItem>
      <FormItem
        label="Department"
        {...formItemLayout}
      >
        {getFieldDecorator('department', {
          rules: [
             { required: true, message: 'Please enter employee\'s department.', whitespace: true },
             { pattern: /^[a-z ]+$/i, message: 'May contain only letters.' },
          ],
        })(<Input />)}
      </FormItem>
      <FormItem {...tailFormItemLayout}>
        <Button tabIndex="0" style={{ float: 'left' }} type="primary" htmlType="submit">Add</Button>
      </FormItem>
    </Form>
  );
});

class WrappedEmployeeForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      success: false,
    };

    this.changeSubmitStatus = this.changeSubmitStatus.bind(this);
  }

  changeSubmitStatus() {
    this.setState({ success: true });
  }
  render() {
    if (this.state.success) {
      return <Redirect to="/" />;
    }
    return (
      <div style={{ width: '50%' }}>
        <Link to="/">
          <Button
            style={{ float: 'left', margin: '10px' }}
            tabIndex="0"
            size="small"
            type="primary"
            icon="left"
          >
            Back
          </Button>
        </Link>
        <h1 style={{ clear: 'left' }}>Add Employee</h1>
        <EmployeeForm
          pageChanger={this.props.setCurrentPage}
          cursorChanger={this.props.setCursor}
          submitStatus={this.state.success}
          statusHandler={this.changeSubmitStatus}
          {...this.props.fields}
          onChange={this.props.updateFields}
          onGoodSubmit={this.props.addEmployee}
        />
      </div>
    );
  }
}

/* eslint-disable react/forbid-prop-types */
WrappedEmployeeForm.propTypes = {
  addEmployee: PropTypes.func.isRequired,
  updateFields: PropTypes.func.isRequired,
  fields: PropTypes.object,
  setCurrentPage: PropTypes.func.isRequired,
  setCursor: PropTypes.func.isRequired,
};
/* eslint-enable */


const mapStateToProps = state => ({
  fields: state.form.fields,
});

const mapDispatchToProps = dispatch => (
  bindActionCreators({
    updateFields,
    addEmployee,
    setCurrentPage,
    setCursor,
  }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(WrappedEmployeeForm);

