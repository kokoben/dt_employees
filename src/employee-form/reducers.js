import { combineReducers } from 'redux';
import * as actions from './actions/types';

export const fields = (state = null, action) => {
  switch (action.type) {
    case actions.FIELDS_UPDATE:
      if (!state) return null;
      return {
        ...state,
        ...action.fields,
      };
    default:
      return state;
  }
};

export default combineReducers({
  fields,
});
