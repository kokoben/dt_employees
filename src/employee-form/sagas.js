import axios from 'axios';
import { put, call, takeEvery } from 'redux-saga/effects';
import * as actions from './actions/types';
import { postEmployee } from './api-calls';

// workers
function* addEmployeeAsync(action) {
  try {
    let name = `${action.fields.lastName}, ${action.fields.firstName}`;
    name = name.toUpperCase();

    const data = JSON.stringify({
      name,
      department: action.fields.department.toUpperCase(),
      employee_annual_salary: action.fields.salary,
      job_titles: action.fields.jobTitle.toUpperCase(),
    });

    const response = yield call(axios.post, postEmployee(), data);
    yield put({ type: actions.EMPLOYEE_ADD_SUCCESS, response });
  } catch (e) {
    console.log('employee_add failed');
    yield put({ type: actions.EMPLOYEE_ADD_FAIL, message: e.message });
  }
}


// watchers
export function* watchAddEmployee() {
  yield takeEvery(actions.EMPLOYEE_ADD, addEmployeeAsync);
}
