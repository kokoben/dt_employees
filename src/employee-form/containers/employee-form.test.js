import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { WrappedEmployeeForm } from './employee-form';

it('renders correctly', () => {
  const mockUpdateFields = jest.fn();
  const mockAddEmployee = jest.fn();
  const mockSetCurrentPage = jest.fn();
  const mockSetCursor = jest.fn();

  const wrapper = shallow(<WrappedEmployeeForm
      addEmployee={mockAddEmployee}
      updateFields={mockUpdateFields}
      fields = {{}}
      setCurrentPage={mockSetCurrentPage}
      setCursor={mockSetCursor}
  />);

      const tree = toJson(wrapper);
      expect(tree).toMatchSnapshot();
});
