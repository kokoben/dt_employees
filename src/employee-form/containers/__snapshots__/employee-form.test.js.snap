// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/"
  >
    <Button
      ghost={false}
      icon="left"
      loading={false}
      prefixCls="ant-btn"
      size="small"
      style={
        Object {
          "float": "left",
          "margin": "10px",
        }
      }
      tabIndex="0"
      type="primary"
    >
      Back
    </Button>
  </Link>
  <div
    className="employee-form"
  >
    <h1
      style={
        Object {
          "clear": "left",
        }
      }
    >
      Add Employee
    </h1>
    <Form(WrappedComponent)
      cursorChanger={[Function]}
      onChange={[Function]}
      onGoodSubmit={[Function]}
      pageChanger={[Function]}
      statusHandler={[Function]}
      submitStatus={false}
    />
  </div>
</div>
`;
